### PR TITLE
8257190: simplify PhaseIdealLoop constructors

### PR DESCRIPTION
**Pull request by navyxliu**
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux additional | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- | ----- |
| Build | ✔️ (8/8 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |  ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ❌ (1/9 failed) |

**Failed test task**
- [macOS x64 (hs/tier1 runtime)](https://github.com/navyxliu/jdk/runs/1476147682)

### Issue
 * [JDK-8257190](https://bugs.openjdk.java.net/browse/JDK-8257190): simplify PhaseIdealLoop constructors

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to 738546fba720bdf8243aa752392ce4f2ed04da33

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1473/head:pull/1473`
`$ git checkout pull/1473`

### CODE REVIEW

**Comment by navyxliu** [comment]
/summary Currently, C2 has 3 private constructors of PhaseIdealLoop as follows. a-b are for verification. only c is for real loop optimizations.
a. PhaseIdealLoop( PhaseIterGVN &igvn)
b. PhaseIdealLoop(PhaseIterGVN &igvn, const PhaseIdealLoop *verify_me)
c. PhaseIdealLoop(PhaseIterGVN &igvn, LoopOptsMode mode)

I propose 3 changes to simplify them.
1. add assertion in the constructor c. C2 shouldn't use mode = LoopOptsVerify for it.
2. merge a and b into one constructor.
3. make the merged verification ctor only for debug builds.

File: src/hotspot/share/opto/loopnode.hpp [hunk 1]
```diff
@@ -931,44 +931,37 @@ class PhaseIdealLoop : public PhaseTransform {
...
     _nodes_required(UINT_MAX) {
+    assert(mode != LoopOptsVerify, "wrong constructor to verify IdealLoop");
     build_and_optimize(mode);
   }
 
+#ifndef PRODUCT
+  // Perform verification that the graph is valid when verify_me is nullptr
   // Verify that verify_me made the same decisions as a fresh run.
-  PhaseIdealLoop(PhaseIterGVN &igvn, const PhaseIdealLoop *verify_me) :
+  PhaseIdealLoop(PhaseIterGVN& igvn, const PhaseIdealLoop* verify_me = nullptr) :
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/loopnode.hpp:955]: I think the comment should be something like:
```
// Verify that verify_me made the same decisions as a fresh run
// or only verify that the graph is valid if verify_me is null.
```

**Review by TobiHartmann — APPROVED** [review]
Otherwise looks good to me.

**Review by vnkozlov — APPROVED** [review]

File: src/hotspot/share/opto/loopnode.hpp (hunk 1 above)
Comment (navyxliu) [reply src/hotspot/share/opto/loopnode.hpp:955]: hi, Tobias, 
thanks for reviewing it. I update the comments.

**Review by TobiHartmann — APPROVED** [review]
Looks good, thanks for updating.

**Comment by navyxliu** [comment]
hi, @TobiHartmann , 

Thank you! Could you sponsor this change? 
The failed test runtime/handshake/HandshakeWalkStackTest.java is a flaky test, which is irrelevant to PR.

thanks,
--lx

**Comment by navyxliu** [comment]
Thank you @DamonFool and other reviewers.

**Pull request closed without merging**